src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,232 @@
+                                           cap.tag.rndv.max_iov),
+       .super.hdr_size      = 0,
+       .super.send_op       = UCT_EP_OP_RNDV_ZCOPY,
+       .super.memtype_op    = UCT_EP_OP_LAST,
+       .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY |
+                              UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY |
+                              UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG,
+       .super.exclude_map   = 0,
+       .lane_type           = UCP_LANE_TYPE_TAG,
+       .tl_cap_flags        = UCT_IFACE_FLAG_GET_ZCOPY |
+                              UCT_IFACE_FLAG_TAG_RNDV_ZCOPY
+    };
+
+    if (!ucp_tag_rndv_check_op_id(init_params) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
+}
+
+static void
+ucp_tag_rndv_offload_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_send_request_id_release(req);
+    ucp_proto_request_zcopy_complete(req, uct_comp->status);
+}
+
+static ucs_status_t
+ucp_tag_rndv_offload_request_init(ucp_request_t *req, ucp_md_map_t md_map,
+                                  uct_completion_callback_t comp_func,
+                                  unsigned uct_reg_flags, unsigned dt_mask)
+{
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    status = ucp_ep_resolve_remote_id(req->send.ep, spriv->super.lane);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assert(md_map != 0);
+    ucs_assertv(dt_mask == UCS_BIT(UCP_DATATYPE_CONTIG), "dt_mask 0x%x",
+                dt_mask);
+
+    status = ucp_proto_request_zcopy_init(req, md_map, comp_func, uct_reg_flags,
+                                          dt_mask);
+    if (status == UCS_OK) {
+        ucp_send_request_id_alloc(req);
+    }
+
+    return status;
+}
+
+static ucs_status_t
+ucp_tag_rndv_offload_send_func(ucp_request_t *req,
+                               const ucp_proto_single_priv_t *spriv,
+                               uct_iov_t *iov)
+{
+    ucp_tag_offload_unexp_rndv_hdr_t rndv_hdr = {
+        .ep_id    = ucp_send_request_get_ep_remote_id(req),
+        .req_id   = ucp_send_request_get_id(req),
+        .md_index = spriv->super.md_index
+    };
+    void *rndv_op;
+
+    rndv_op = uct_ep_tag_rndv_zcopy(ucp_ep_get_fast_lane(req->send.ep,
+                                                         spriv->super.lane),
+                                    req->send.msg_proto.tag,
+                                    &rndv_hdr, sizeof(rndv_hdr), iov, 1, 0,
+                                    &req->send.state.uct_comp);
+
+    if (ucs_unlikely(UCS_PTR_IS_ERR(rndv_op))) {
+        return UCS_PTR_STATUS(rndv_op);
+    }
+
+    req->flags                   |= UCP_REQUEST_FLAG_OFFLOADED;
```

Comment:
maybe it should be part of ucp_tag_rndv_offload_request_init ?

Reply:
i'd keep it here, because it means that rndv operation was sucesfully sent thru UCT TAG API and we need to cancel it there.

Reply:
but when we call ucp_tag_rndv_offload_request_init, we already know the request is going to be sent thru tag api only. and that function should initialize the fields that do not change during protocol progress

Reply:
i'd want to keep it here, because this flag shows that `req->send.tag_offload.rndv_op` is initialized and therefore needs to be freed/canceled in some cleanup flow. Setting this flag earlier (before the actual operation is peformed) is error prone, because it may lead to incorect flows during cleaup/error scenarios.
UCT API does not define invalid rndv operation id, so we have to maintain this flag to check whether UCT rndv resource needs to be freed.